### PR TITLE
fix(notification): 修复总开关关闭后排行榜/成本预警仍推送 & 成本预警间隔≥60分钟塌缩为每小时 (#1236)

### PR DESCRIPTION
**Pull request by ding113**
## 背景

修复 #1236：用户反馈「每日用户消费排行榜」推送异常 ——

1. 设为每日 17:00 执行一次正常；
2. 改成 18:00 后变成每隔 1 小时执行；
3. 关闭通知总开关后，仍每隔 1 小时推送到企业微信。

## 根因分析

经多代理交叉排查（4 个并行调查 + 3 个对抗式验证 agent），定位到以下问题：

### 1. 执行期未复核开关（症状 3 的直接原因）
`notification-queue.ts` 队列处理器中，`daily-leaderboard` 与 `cost-alert` 两个 case **不会在任务执行时复核** `settings.enabled` / 子开关，只判断是否有数据。这与 `cache-hit-rate-alert`(`:267`) 及独立函数 `notifier.ts:139` 的行为不一致。因此当总开关关闭后，关闭前已入队的 repeatable 作业仍会继续触发并发送。

### 2. 仅生产环境重新调度（症状 2/3 的放大器）
`updateNotificationSettingsAction` 仅在 `NODE_ENV === "production"` 时调用 `scheduleNotifications()`，而默认值为 `"development"`。未显式设置该变量的自部署环境在改设置（含关闭总开关）时**不会刷新/移除 Bull repeatable 作业**，导致旧调度残留、关开关无效。`scheduleNotifications` 内部已 fail-open，启动入口与 `notification-bindings` 早已无条件调用它，移除该 gate 安全且一致。

### 3. 成本预警间隔 ≥60 分钟塌缩为每小时（真正的「每隔 1 小时」来源）
`cost-alert` 使用 `*/${interval} * * * *`，而 Bull 的 cron 分钟字段仅 0-59，`*/60` 会塌缩成「每小时第 0 分」，`*/120` 同样退化为每小时。默认 `costAlertCheckInterval=60` 即命中此 bug。`cache-hit-rate-alert` 早已用 `<=59 ? cron : every` 规避，此次让成本预警对齐该逻辑（legacy + targets 两条路径）。

## 改动

- **执行期复核开关**：`daily-leaderboard`、`cost-alert` 处理器在生成/发送前校验 `enabled` 与对应子开关，关闭即 `skipped`。
- **移除 NODE_ENV gate**：设置变更后始终重新调度，使开关/时间/间隔变更即时生效（添加/移除 repeatable 作业）。
- **成本预警间隔修复**：间隔 ≥60 分钟改用 `{ every }`，legacy 与 targets 模式一致；间隔做 `[1, 1440]` 归一化。
- **重调度健壮性**：抽取 `removeAllRepeatableJobs`，单个 key 移除失败（Redis 瞬时错误）不再中断整轮重调度。

## 测试

新增 `tests/unit/notification/notification-queue.test.ts`（9 个用例）：
- 总开关 / 子开关关闭时处理器 `skipped`，且不调用生成与发送；
- 开关全开时正常发送；
- 总开关关闭时 `scheduleNotifications` 移除全部 repeatable 作业；
- 单个移除失败不影响整体；
- 成本预警 interval=60 → `{ every: 3600000 }`，interval=30 → `*/30 * * * *`。

## 本地验证（全绿）

`build` ✅ / `lint` ✅ / `typecheck` ✅ / `test` ✅（6180 passed, 13 skipped，含本次新增 9 项）

## 范围说明

- 未对 per-binding 自定义 `scheduleCron` 加校验：`0 * * * *` 是合法 cron，属用户意图，非 bug。
- `circuit-breaker` 为事件驱动而非定时任务，无需此修复。

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details open><summary><h3>Greptile Summary</h3></summary>

This PR fixes three root causes behind #1236: (1) the queue processor for `daily-leaderboard` and `cost-alert` now re-checks `settings.enabled` / sub-switches at execution time so stale repeatable jobs can no longer fire after the master switch is turned off; (2) the `NODE_ENV === \"production\"` gate that silently prevented non-production deployments from rescheduling Bull jobs on settings changes is removed; (3) `*/60 * * * *` cron collapse is fixed by routing intervals ≥60 min through `{ every: ms }`.

- **Execution-time guard**: `circuit-breaker`, `daily-leaderboard`, and `cost-alert` processor branches all early-return `{ success: true, skipped: true }` when the relevant switches are off.
- **Interval routing**: new `intervalToRepeat(n, tz?)` helper centralises the `cron`-vs-`every` decision; `clampIntervalMinutes` normalises raw DB values to `[1, 1440]`.
- **Robustness**: `removeAllRepeatableJobs` tolerates single-key Redis failures; if any removal fails during rescheduling, `scheduleNotifications` aborts rather than creating duplicate jobs.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — all three documented bug scenarios are closed, the helpers are correct for all edge cases, and the added test suite exercises both the skip paths and the interval-routing decisions end-to-end.

The changes are narrowly scoped to the notification queue: adding execution-time guards, removing the environment gate that made settings changes a no-op in most self-hosted deployments, and fixing the cron collapse. No auth, data-model, or external API surfaces are touched. The intervalToRepeat helper handles boundary values correctly, and removeAllRepeatableJobs gracefully tolerates partial Redis failures without risking duplicate job scheduling.

No files require special attention.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/lib/notification/notification-queue.ts | Core fix: adds execution-time settings guards to three processor cases, extracts removeAllRepeatableJobs / clampIntervalMinutes / intervalToRepeat helpers, and correctly routes intervals to cron or every. Edge cases handled correctly. |
| src/actions/notifications.ts | Removes NODE_ENV === production guard so scheduleNotifications() is always called after a settings update. |
| tests/unit/notification/notification-queue.test.ts | New test file with 15 cases covering master/sub-switch skipping, fault-tolerant removal, abort-on-partial-remove, and both legacy and targets-mode cron/every routing. |

</details>

</details>

<sub>Reviews (3): Last reviewed commit: ["test(notification): add circuit-breaker ..."](https://github.com/ding113/claude-code-hub/commit/d42200d98b5b4e0303a8b722e05a60180d0d4063) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=35113049)</sub>

### CODE REVIEW

File: src/lib/notification/notification-queue.ts [hunk 1]
```diff
@@ -764,8 +791,11 @@ export async function scheduleNotifications() {
       }
 
       if (settings.costAlertEnabled && settings.costAlertWebhook) {
-        const interval = settings.costAlertCheckInterval ?? 60; // 分钟
-        const cron = `*/${interval} * * * *`; // 每 N 分钟
+        const intervalRaw = settings.costAlertCheckInterval ?? 60; // 分钟
+        const interval = Math.min(Math.max(1, Math.trunc(intervalRaw)), 24 * 60);
+        // Bull cron 分钟字段只有 0-59，`*/60` 会塌缩成“每小时第 0 分”，所以 >=60 分钟改用固定间隔
+        const repeat =
+          interval <= 59 ? { cron: `*/${interval} * * * *` } : { every: interval * 60 * 1000 };
```
[the remark this answers is not in the document]
Comment (ding113) [reply src/lib/notification/notification-queue.ts]: 已采纳。在 545b98f5 提取了 `intervalToRepeat()`：仅当 N<=59 且 60 % N === 0 时使用 cron，否则退化为 { every }，避免 45 等不整除 60 的间隔在整点边界（:45→:00）出现不均匀触发。legacy 与 targets 两处已统一走该 helper。

File: src/lib/notification/notification-queue.ts [hunk 2]
```diff
@@ -848,12 +878,18 @@ export async function scheduleNotifications() {
...
         for (const binding of bindings) {
-          const cron = binding.scheduleCron ?? defaultCron;
           const tz = binding.scheduleTimezone ?? systemTimezone;
+          // 优先级：绑定自定义 cron（支持 tz）> 默认间隔。
+          // Bull cron 分钟字段只有 0-59，默认间隔 >=60 分钟改用固定 every（不支持 tz）。
+          const repeat = binding.scheduleCron
+            ? { cron: binding.scheduleCron, tz }
+            : interval <= 59
+              ? { cron: `*/${interval} * * * *`, tz }
+              : { every: interval * 60 * 1000 };
```
[the remark this answers is not in the document]
Comment (ding113) [reply src/lib/notification/notification-queue.ts]: 已采纳。targets 模式同样改为经由 `intervalToRepeat(interval, tz)`：N 能整除 60 时用 cron（携带 tz），否则用 { every }。见 545b98f5。

File: src/lib/notification/notification-queue.ts [hunk 3]
```diff
@@ -705,6 +719,25 @@ export async function addNotificationJobForTarget(
...
+/**
+ * 移除队列中所有 repeatable 定时任务。
+ * 单个 key 移除失败（如 Redis 瞬时错误）不应中断整个重调度流程，否则会残留旧任务。
+ */
+async function removeAllRepeatableJobs(queue: Queue.Queue<NotificationJobData>): Promise<void> {
+  const repeatableJobs = await queue.getRepeatableJobs();
+  for (const job of repeatableJobs) {
+    try {
+      await queue.removeRepeatableByKey(job.key);
+    } catch (error) {
```
[the remark this answers is not in the document]
Comment (ding113) [reply src/lib/notification/notification-queue.ts]: 已采纳。removeAllRepeatableJobs 现返回 boolean（仍逐个尝试移除全部 key，单个失败不中断）；启用态重调度时若有旧 key 未能移除则中止新增任务并记录 schedule_notifications_aborted，避免新旧 repeatable 同时触发。已加单测 'aborts adding new jobs when an old repeatable cannot be removed'。

File: src/lib/notification/notification-queue.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ding113) [reply src/lib/notification/notification-queue.ts]: 说明：every 按固定毫秒节奏触发、不对齐墙钟，时区对纯固定间隔本身无意义，故仅 cron 路径携带 tz。调度日志的 schedule 字段会输出 cron 表达式或 every:Nm，可据此观测实际路径；此为 Bull 固有限制，cache-hit-rate 既有逻辑一致。

File: tests/unit/notification/notification-queue.test.ts [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+    queueRemoveRepeatableByKey.mockRejectedValueOnce(new Error("redis down"));
+
+    const { scheduleNotifications } = await import("@/lib/notification/notification-queue");
+    await expect(scheduleNotifications()).resolves.toBeUndefined();
+
+    expect(queueRemoveRepeatableByKey).toHaveBeenCalledTimes(2);
+  });
+
+  it("uses {every} instead of */60 cron for cost-alert interval >= 60 (legacy)", async () => {
+    mockGetNotificationSettings.mockResolvedValue(
```
[the remark this answers is not in the document]
Comment (ding113) [reply tests/unit/notification/notification-queue.test.ts:418]: 已采纳。新增 targets 模式 cost-alert 调度单测：校验 binding 专属 jobId（cost-alert:7）、cron 路径携带 tz（Asia/Tokyo）、以及 interval>=60 走 { every }。见 545b98f5。

**Comment by ding113** [comment]
## 已处理全部 review 反馈(commit 545b98f5)

| 来源 | 反馈 | 处理 |
|------|------|------|
| gemini (medium) | `*/N` 在 N 不整除 60 时整点边界间隔不均 | 提取 `intervalToRepeat()`,仅当 `N<=59 && 60 % N === 0` 用 cron,否则 `{ every }` |
| codex (P2) | 旧 repeatable 移除失败后仍新增新任务 → 双重触发 | `removeAllRepeatableJobs` 返回 boolean;启用态移除失败则中止新增并记 `schedule_notifications_aborted` |
| greptile (P2) | `{ every }` 丢失 tz | 说明:固定间隔对 tz 无意义,`schedule` 日志字段已可观测路径;Bull 固有限制 |
| coderabbit / greptile | 缺正向 cost-alert 用例、targets 模式未覆盖 | 新增正向 cost-alert、targets 模式(jobId+tz+every)、circuit-breaker、移除失败中止等用例 |
| 自检 (multi-agent review) | circuit-breaker 处理器缺执行期开关校验;4 处 interval→repeat 重复 | 补 circuit-breaker 执行期校验;抽取 `clampIntervalMinutes`/`intervalToRepeat` 统一 4 处 |

本地全绿:`build` / `lint` / `typecheck` / `test`(单测 6200 项,新增通知队列用例 16 项)。
